xray/core/rolling.py

```diff
@@ -0,0 +1,238 @@
+import numpy as np
+import bottleneck as bn
+
+from .pycompat import OrderedDict, zip
+
+from .common import ImplementsArrayReduce
+
+
+class Rolling(object):
+    """A object that implements the moving window pattern.
+
+    See Also
+    --------
+    Dataset.groupby
+    DataArray.groupby
+    Dataset.rolling
+    DataArray.rolling
+    """
+
+    _attributes = ['window', 'min_periods', 'center', 'dim']
+
+    def __init__(self, obj, min_periods=None, center=False, **kwarg):
+        """
+        Moving window object.
+
+        Parameters
+        ----------
+        obj : Dataset or DataArray
+            Object to window.
+        min_periods : int, default None
+            Minimum number of observations in window required to have a value
+            (otherwise result is NA). The default, None, is equivalent to
+            setting min_periods equal to the size of the window.
+        center : boolean, default False
+            Set the labels at the center of the window.
+        kwarg : dim=window
+            dim : str
+                Name of the dimension to create the rolling iterator
+                along (e.g., `time`).
+            window : int
+                Size of the moving window.
+
+        Returns
+        -------
+        rolling : type of input argument
+        """
+
+        if len(kwarg) != 1:
+            raise ValueError('only one dim/window can be provided')
+
+        dim, window = next(iter(kwarg.items()))
+
+        if window <= 0:
+            raise ValueError('window must be > 0')
+
+        self.obj = obj
+
+        # attributes
+        self.window = window
+        self.min_periods = min_periods
+        self.center = center
+        self.dim = dim
+
+        self._axis_num = self.obj.get_axis_num(self.dim)
+
+        self._windows = None
+        self._valid_windows = None
+        self.window_indices = None
+        self.window_labels = None
+
+        self._setup_windows(min_periods=min_periods, center=center)
+
+    def __unicode__(self):
```

Comment:
yeah, i was wondering the same thing.  As you can see, I just copied this from the Pandas PR.